crates/pico_macros/src/memo.rs

```diff
@@ -97,56 +90,97 @@ pub(crate) fn memo(args: TokenStream, item: TokenStream) -> TokenStream {
 
     let mut new_sig = sig.clone();
 
-    if args_.reference {
-        let lifetime = new_sig.generics.params.iter().find_map(|param| {
-            if let GenericParam::Lifetime(lt) = param {
-                Some(&lt.lifetime)
-            } else {
-                None
-            }
-        });
+    let macro_args = match parse_macro_args(args, &sig) {
+        Ok(parsed) => parsed,
+        Err(e) => return e.to_compile_error().into(),
+    };
 
-        if let Some(lt) = lifetime {
-            new_sig.output =
-                ReturnType::Type(parse_quote!(->), Box::new(parse_quote!(&#lt #return_type)));
+    if macro_args.inner {
+        return_expr = quote! { #return_expr.clone() };
+    } else {
+        let lifetime = get_fn_lifetime(&mut new_sig);
+        if let Err(e) = check_db_lifetime(&mut new_sig.inputs[0], lifetime.clone()) {
+            return e.to_compile_error().into();
+        }
+        if macro_args.inner_ref {
+            new_sig.output = ReturnType::Type(
+                parse_quote!(->),
+                Box::new(parse_quote!(&#lifetime #return_type)),
+            );
         } else {
-            new_sig.generics.params.push(parse_quote!('db));
-            if let FnArg::Typed(PatType { ty, .. }) = &mut new_sig.inputs[0] {
-                if let syn::Type::Reference(ref mut reference) = **ty {
-                    reference.lifetime = Some(parse_quote!('db));
-                } else {
-                    return Error::new_spanned(ty, "Expected a mutable reference type")
-                        .to_compile_error()
-                        .into();
-                }
-            }
-            new_sig.output =
-                ReturnType::Type(parse_quote!(->), Box::new(parse_quote!(&'db #return_type)));
+            new_sig.output = ReturnType::Type(
+                parse_quote!(->),
+                Box::new(parse_quote!(::pico::MemoRef<#lifetime, #return_type>)),
+            );
+            return_expr = quote! { ::pico::MemoRef::new(#db_arg, derived_node_id) };
         }
-    } else {
-        return_expr = quote! { #return_expr.clone() };
     }
 
-    let unpacked_args = other_args.clone();
-    let inner_args = other_args.clone();
+    let extract_parameters = other_args.clone().zip(argument_types.clone())
+        .enumerate()
+        .map(|(i, (arg, ty))| {
+            match ArgType::parse(ty) {
+                ArgType::Source => {
+                    let maybe_ref = if matches!(**ty, syn::Type::Reference(_)) {
```

Comment:
tbh it doesn't make sense to ever pass a reference to a `Source` or `MemoRef`, since they implement `Copy`

This is more of a comment than a request to change anything

Reply:
Agreed, but what should we do? Accept references or throw a compilation error?

Reply:
accept haha, just talking to myself. Maybe warn in debug mode?